api/base/parsers.py

```diff
@@ -15,30 +18,62 @@ class JSONAPIParser(JSONParser):
     media_type = 'application/vnd.api+json'
     renderer_class = JSONAPIRenderer
 
+    @staticmethod
+    def flatten_data(resource_object, request_method, is_list=False):
+        """
+        Flattens data objects, making attributes fields the same level as id and type.
+        """
+
+        if "attributes" not in resource_object and request_method != 'DELETE':
+            raise JSONAPIException(source={'pointer': '/data/attributes'}, detail=NO_ATTRIBUTES_ERROR)
+
+        object_id = resource_object.get('id')
+        object_type = resource_object.get('type')
```

Comment:
We created an ID and a TypeField which validate later on if they're missing for singular requests.  I ensure bulk object_id and bulk type are present here because of the all-or-nothing rules.  For bulk requests, if any of the id's or types are missing, we want to throw an error now because every action needs to succeed or they all should fail.  For singular requests, I can wait until the serializer validation.